test/src/main/java/org/apache/accumulo/test/ScanServerShutdownIT.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.fate.zookeeper.ZooReaderWriter;
+import org.apache.accumulo.core.security.Authorizations;
+import org.apache.accumulo.core.spi.scan.ScanServerSelector;
+import org.apache.accumulo.harness.MiniClusterConfigurationCallback;
+import org.apache.accumulo.harness.SharedMiniClusterBase;
+import org.apache.accumulo.minicluster.ServerType;
+import org.apache.accumulo.miniclusterImpl.MiniAccumuloConfigImpl;
+import org.apache.accumulo.server.ServerContext;
+import org.apache.accumulo.test.util.Wait;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class ScanServerShutdownIT extends SharedMiniClusterBase {
+
+  private static class ScanServerShutdownITConfiguration
+      implements MiniClusterConfigurationCallback {
+
+    @Override
+    public void configureMiniCluster(MiniAccumuloConfigImpl cfg,
+        org.apache.hadoop.conf.Configuration coreSite) {
+      cfg.setNumScanServers(1);
+
+      // Timeout scan sessions after being idle for 3 seconds
+      cfg.setProperty(Property.TSERV_SESSION_MAXIDLE, "3s");
+
+      // Configure the scan server to only have 1 scan executor thread. This means
+      // that the scan server will run scans serially, not concurrently.
+      cfg.setProperty(Property.SSERV_SCAN_EXECUTORS_DEFAULT_THREADS, "1");
+    }
+  }
+
+  @BeforeAll
+  public static void start() throws Exception {
+    ScanServerShutdownITConfiguration c = new ScanServerShutdownITConfiguration();
+    SharedMiniClusterBase.startMiniClusterWithConfig(c);
+  }
+
+  @AfterAll
+  public static void stop() throws Exception {
+    SharedMiniClusterBase.stopMiniCluster();
+  }
+
+  @Test
+  public void testRefRemovalOnShutdown() throws Exception {
+
+    ServerContext ctx = getCluster().getServerContext();
+    String zooRoot = ctx.getZooKeeperRoot();
+    ZooReaderWriter zrw = ctx.getZooReaderWriter();
+    String scanServerRoot = zooRoot + Constants.ZSSERVERS;
+
+    Wait.waitFor(() -> zrw.getChildren(scanServerRoot).size() == 0);
+
+    // Stop normal ScanServers so that we can start our custom implementation
+    // that shuts down after 3 batch scans
+    getCluster().getClusterControl().startScanServer(SelfStoppingScanServer.class, 1,
+        ScanServerSelector.DEFAULT_SCAN_SERVER_GROUP_NAME);
+
+    // Wait for the ScanServer to register in ZK
+    Wait.waitFor(() -> zrw.getChildren(scanServerRoot).size() == 1);
+
+    try (AccumuloClient client = Accumulo.newClient().from(getClientProps()).build()) {
+      final String tableName = getUniqueNames(1)[0];
+
+      client.tableOperations().create(tableName);
+
+      // Make multiple files
+      final int fileCount = 3;
+      for (int i = 0; i < fileCount; i++) {
+        ScanServerIT.ingest(client, tableName, 10, 10, 0, "colf", true);
+      }
+
+      for (int i = 0; i < 3; i++) {
```

Comment:
On the 2nd iteration could verify that some refs exists.

```suggestion
      for (int i = 0; i < 3; i++) {
         if(i ==1){
             // the 1st scan should have created some refs
             assertTrue(ctx.getAmple().getScanServerFileReferences().count()>0);
         }
```

Reply:
Do they get cleaned up when the scan closes? If not, then the check on line 120 below should fail, right? It should be 3, 6, 9 and not 3 every time?

Reply:
Yeah the check on line 120 is sufficient, I missed that.   I think I glossed over that code block assuming it was just scanning.